---
title: "On making and loading NestLink data"
author:
- name: Christian Panse
  affiliation: Functional Genomics Center Zurich, UZH|ETHZ
  email: cp@fgcz.ethz.ch
date: '2018-12-13'
output:
  BiocStyle::html_document
bibliography:
  - NestLink.bib
abstract: |
    Mass spectrometry data are available via ProteomeXchange with identifier
    [PXD009301](https://www.ebi.ac.uk/pride/archive/login). 
    NGS datasets are available on the European Nucleotide Archive (ENA)
    under accession number 
    [PRJEB25673](https://www.ebi.ac.uk/ena/data/search?query=PRJEB25673). 
    The NGS and MS data were handled and
    annotated using the [B-Fabric](https://fgcz-bfabric.uzh.ch) 
    information management system and are available
    for registered users under project identifiers 1644 and 1875.
vignette: |
  %\VignetteIndexEntry{On loading NestLink data using ExperimentHub}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
# Make Data (replaces `make-data.R`)

## `NL42_100K.fastq.gz`
TODO(LO)

```{r eval=FALSE}
NL42_100K <- NestLink:::.getReadsFromFastq("inst/extdata/NL42_100K.fastq.gz")
save(NL42_100K, file="inst/extdata/NestLink_NL42_100K.RData")
```

## `knownNB.txt`
TODO(LO)


## `nanobodyFlycodeLinkage.RData`
NGS ground truth derived by appling the function `runNGSAnalysis`
to the previous files.

## `NB.tryptic` and `FC.tryptic`

* https://fgcz-gstore.uzh.ch/projects/p1644/analysis_20170609_o3040/p1644o3482-4_S4.extendedFrags_uniqNB2FC.txt
* https://fgcz-gstore.uzh.ch/projects/p1644/analysis_20170609_o3040/p1644o3482-5_S5.extendedFrags_uniqNB2FC.txt

## `F255744.RData` and `WU160118.RData`

the mass spec files below availbale through [ProteomeXchange PXD009301](https://www.ebi.ac.uk/pride/archive/login) 

```{r echo=FALSE, message=FALSE, warning=FALSE}
library(knitr)
library(ExperimentHub)
eh <- ExperimentHub(); 
load(query(eh, c("NestLink", "WU160118.RData"))[[1]])
class(WU160118)
PATTERN <- "^GS[ASTNQDEFVLYWGP]{7}(WR|WLTVR|WQEGGR|WLR|WQSR)$"
idx <- grepl(PATTERN, WU160118$pep_seq)
WU <- WU160118[idx & WU160118$pep_score > 25,]

kable(unique(unlist(lapply(strsplit(x=as.character(WU$pep_scan_title), split=";"), function(x)gsub("File:", "", gsub("\\\\", "/", x[1]))))))

```

were assigned to peptide sequences using the parameter listed in the table below and the Matrix Science Mascot Server version 2.5. 

|Parameter|Value|
|:---|:---|
|COM|170819_MS1708116_NL5idx4to5_Competition2BG_db8_db10_swissprot_d_merge|
|TOL|10|
|TOLU|ppm|
|ITOL|0.6| 
|ITOLU|Da|
|USERNAME|egloffp|
|CHARGE|2+|
|IT_MODS|Deamidated (NQ),Oxidation (M)|
|INSTRUMENT|ESI-TRAP|
|release|fgcz_swissprot_d_20140403.fasta|

The results were exported as XML.
The XML was parsed and exported as data.frame using `r CRANpkg('protViz')` function `protViz:::as.data.frame.mascot`.




## `PGexport2_normalizedAgainstSBstandards_Peptides.csv`

* Workunit : 158716 - QEXACTIVEHF_1

    * `20170919_16_62465_nl5idx1-3_6titratecoli.raw`
    * `20170919_05_62465_nl5idx1-3_6titratecoli.raw`

* Workunit : 158717 - QEXACTIVEHF_1

    * `20170919_14_62466_nl5idx1-3_7titratesmeg.raw`
    * `20170919_09_62466_nl5idx1-3_7titratesmeg.raw`

Two LC-MS/MS runs were aligned in Progenesis QI (Nonlinear Dynamics) with an
alignment score of 93.1 %, followed by peak picking with an allowed ion
charge of +2 to +5. 

# Uploading to S3
```
#!/bin/bash

aws --profile AnnotationContributor s3 cp NestLink/F255744.RData s3://annotation-contributor/NestLink/F255744.RData --acl public-read

aws --profile AnnotationContributor s3 cp NestLink/WU160118.RData s3://annotation-contributor/NestLink/WU160118.RData --acl public-read

aws --profile AnnotationContributor s3 cp NestLink s3://annotation-contributor/NestLink --recursive --acl public-read

```
# Getting started using  Bioconductor ExperimentHub

load metadata

```{r eval=TRUE}

fl <- system.file("extdata", "metadata.csv", package='NestLink')
kable(metadata <- read.csv(fl, stringsAsFactors=FALSE))
```

load `r Biocpkg("NestLink")` package data from aws s3

```{r echo=TRUE, message=FALSE, warning=FALSE, eval=TRUE}
library(ExperimentHub)

eh <- ExperimentHub(); 
query(eh, "NestLink")

load(query(eh, c("NestLink", "F255744.RData"))[[1]])
dim(F255744)

load(query(eh, c("NestLink", "WU160118.RData"))[[1]])
dim(WU160118)
```
